onnx/onnx.in.proto

```diff
@@ -196,6 +204,117 @@ message NodeProto {
   optional string doc_string = 6;
 }
 
+// Training information
+// TrainingInfoProto stores information for training a model.
+// In particular, this defines two functionalities: an initialization-step
+// and a training-algorithm-step. The semantics of the initialization-step is
+// that all initializers (in ModelProto.graph and in all
+// TrainingInfoProto.algorithm) are first initialized as specified by the
+// initializers in the graph, and then by the "initialization_binding" in every
+// instance in ModelProto.training_info. The semantics of the
+// training-algorithm-step is that all the initializers are updated as
+// specified by the "update_binding" in every instance in training_info.
+//
+// The field "algorithm" defines a computation graph similar to the inference
+// one (i.e., ModelProto.graph) but represents a training algorithm's step.
+// Each execution of a TrainingInfoProto.algorithm leads to the update of some
+// initializers in its TrainingInfoProto.algorithm.initializer and
+// ModelProto.graph.initializer.
+//
+// If there are multiple TrainingInfoProto's, sequentially executing all of
+// them once constitutes a full training algorithm. The initializer update
+// happens at the end of each TrainingInfoProto's execution. If the targeted
+// training algorithm contains consecutive update stages (such as block
+// coordinate descent methods), the user needs to create a TrainingInfoProto
+// for each stage.
+//
+// In addition to training (improving the model based on input data),
+// TrainingInfoProto can store an initialization process (resetting the model
+// back to its original state as if no training has been consumed) as a
+// computation graph in the "initialization" field. The execution of
+// "initialization" is similar to that of "algorithm". The user can use
+// "initialization_binding" to assign the outputs of "initialization" to
+// ModelProto.graph.initializer and TrainingInfoProto.algorithm.initializer.
+// 
+// Several inference components are usable in TrainingInfoProto; the list
+// includes
+//   1. ModelProto.graph is a callable inside TrainingInfoProto.algorithm.
+//      Each call is a NodeProto in TrainingInfoProto.algorithm.node with
+//      NodeProto.op_type == "Graph" and
+//      NodeProto.name == ModelProto.graph.name.
+//   2. Initializers in ModelProto.graph.initializer can be referenced by
+//      nodes in TrainingInfoProto.algorithm.nodes and
+//      TrainingInfoProto.initialization.nodes. It implies that training
+//      nodes cannot produce values with names identical to inference
+//      initializers.
```

Comment:
[IR.md](https://github.com/onnx/onnx/blob/master/docs/IR.md#nodes) says
> The graph MUST use single static assignment for all node outputs, which means that all node output names MUST be unique within a graph. In the case of a nested subgraph, a node output name MUST be distinct from the names from the outer scopes that are visible in the nested subgraph.

So I think you could be more explicit -- I think it would be in the spirit of the spec thus far to say value names must be unique in `ModelProto.graph`, `TrainingInfoProto.initialization.nodes`, and `TrainingInfoProto.algorithm.nodes`.

Reply:
What does it mean? When `ModelProto.graph` is called in `TrainingInfoProto.algorithm`, its inputs and outputs would be mapped by position like how a FunctionProto is called. Thus, `ModelProto.graph` and `TrainingInfoProto.algorithm` can have `X, Y ---> Add ---> Z` at the same time as long as the value names in `TrainingInfoProto.algorithm` doesn't conflict with the `initializer` names in `ModelProto.graph.initializer`.

Reply:
Fair enough. I misread that part of the spec that I quoted to mean you could uniquely identify a value in the entire graph, but I don't think that's actually the case, so I don't think my suggestion really makes sense anyway.